test/broadway_rabbitmq/producer_test.exs

```diff
@@ -371,6 +371,7 @@ defmodule BroadwayRabbitMQ.ProducerTest do
     connect_responses = Keyword.get(opts, :connect_responses, [])
     backoff_type = Keyword.get(opts, :backoff_type, :exp)
     requeue = Keyword.get(opts, :requeue, :always)
+    metadata = []
```

Comment:
We also need to test that we're actually retrieving the information from the `meta` map and appending it to our `metadata`. One way we can do this is to receive a custom metadata option:

instead of

```elixir
metadata = []
```

we could have:

```elixir
metadata = Keyword.get(opts, :metadata, [])
```
and then change our fake `deliver_messages` implementation: https://github.com/plataformatec/broadway_rabbitmq/blob/eb4364ed3c61d667d78c4dc33eeceb2c7cbc7196/test/broadway_rabbitmq/producer_test.exs#L415-L422

to something like:

```elixir
defp deliver_messages(broadway, messages, opts \\ []) do
    redelivered = Keyword.get(opts, :redelivered, false)
    producer = Broadway.Server.get_random_producer(broadway)
    extra_metadata = Keyword.get(opts, extra_metadata, %{})
    metadata = Map.merge(%{delivery_tag: msg, redelivered: redelivered}, extra_metadata)

    Enum.each(messages, fn msg ->
      send(producer, {:basic_deliver, msg, metadata})
    end)
  end
```

You'll also need to change the following line, so we can send the whole message instead of the just the `message.data`:

https://github.com/plataformatec/broadway_rabbitmq/blob/eb4364ed3c61d667d78c4dc33eeceb2c7cbc7196/test/broadway_rabbitmq/producer_test.exs#L102

> Note: With this change, we'll have to adjust a couple of tests.

Now we can create a test where we:
  * Fake the received metadata
  * Use our new `metadata` option specifying which information we want
  * Assert if the received message contains (or not) the information we requested

Reply:
I have to admit that I don't totally understand - yet - the way all these tests work, so I'm not sure if what I'm doing is correct or even relevant.
I don't know how to fake the received metadata and check if the received message contains the information requested.

I've already improved a bit - well at least I think so - the existing tests based upon what you've said above.